nondep_type_decl: expand to private abbrev instead of abstracting

trefis opened the pull request:
Given:
```ocaml
# module F(_ : sig end) = struct type t = private int end
  module G(X : sig end) : sig type t = F(X).t end = F(X);;
module F : sig  end -> sig type t = private int end
module G : functor (X : sig  end) -> sig type t = F(X).t end
```
Before this PR:
```ocaml
# module M = G(struct end);;
module M : sig type t end
```

After:
```ocaml
# module M = G(struct end);;
module M : sig type t = private int end
```

trefis committed:
nondep_type_rec: allow expanding to private abbrev

trefis committed:
nondep_type_decl: add a test showing private abbrevs are ignored

trefis committed:
nondep_type_decl: add tests ensuring we don't unroll abbrevs in private case

trefis committed:
Changes

garrigue commented on testsuite/tests/typing-modules/nondep_private_abbrev.ml:
```diff
@@ -0,0 +1,85 @@
+[%%expect{|
+module IndirectPriv : sig type t = private [ `Foo of 'a ] as 'a end
+|}]
+
```
The code looks fine, but I'm concerned by `IndirectPriv`.
What would happen if the original type definition in `Priv` were
```
type t = private [`Foo of t -> int | `Bar of int]
```
Private abbreviations can be soundly expanded by subtyping, but only in covariant positions.

garrigue replied:
I think that you need to track variance in `nondep_type_rec`, and only call `expand_abbrev_opt` at covariant positions.

trefis replied:
Good catch. I do check `is_covariant` before calling `nondep_type_rec` with `~expand_private:true`, but I am indeed recursing without being careful about this.

I'll push a fix.

garrigue commented:
Thinking more, I'm not even 100% sure that this is sound when restricted to covariant positions.
It certainly looks OK, but I would like to see a formal argument.

trefis committed:
nondep_type_rec: add broken test suggested by jacques

trefis committed:
nondep_type_rec: only call expand_abbrev_opt at the toplevel

trefis commented:
I pushed a fix for the issue you pointed to above.

As for restricting to covariant positions: the restriction which is now implemented is stronger than this: we call `expand_abbrev_opt` only at the toplevel, not when recursing.
Doing so we match what the user can already do manually in other parts of the system, e.g. below `s` is in a covariant position, but the user is not allowed to use its expansion:
```
# type s = private int;;                                                           
type s = private int
# module M : sig type t = private int list end = struct type t = s list end;;
Characters 47-73:
  module M : sig type t = private int list end = struct type t = s list end;;
                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: Signature mismatch:
       Modules do not match:
         sig type t = s list end
       is not included in
         sig type t = private int list end
       Type declarations do not match:
         type t = s list
       is not included in
         type t = private int list
# module M : sig type t end = struct type t = s list end;;                   
module M : sig type t end
# 
```

Likewise:
```
# module F(_ : sig end) = struct type t = private int end;;
module F : sig  end -> sig type t = private int end
# module G(X : sig end) = struct type t = F(X).t list end;;
module G : functor (X : sig  end) -> sig type t = F(X).t list end
# module M = G(struct end);;
module M : sig type t end
# 
```

trefis commented:
And to be somewhat more explicit: we allow ourselves the expansion at the toplevel for `nondep_type_rec`, as it is already allowed here:
```
# type s = private int;;
type s = private int
# module M : sig type t = private int end = struct type t = s end;;
module M : sig type t = private int end
```

Which is similar to the example in my initial message:
```
# module F(_ : sig end) = struct type t = private int end
  module G(X : sig end) = F(X);;
module F : sig  end -> sig type t = private int end
module G : functor (X : sig  end) -> sig type t = F(X).t end
# module M = G(struct end);;
module M : sig type t = private int end
```

garrigue commented:
Indeed in `Includecore.type_manifest` we allow expanding `ty1` with `try_expand_opt` when `priv2 = Private`.
So morally what you do looks correct.
However, one might be even more careful, and add an assertion checking the subtyping between the original module type and the result of `Mtype.nondep_supertype`. This would have caught the variance problem. The cost is not significant, so maybe it's time to be more defensive when possible (unfortunately this is not often the case...)

trefis committed:
check subtyping between original mty and result of nondep_supertype

trefis commented:
I added the check you suggested (in commit b5d4f87), it indeed would have caught the variance issue.

I also rebased and updated the changes entry; so it's ready to merge if you're happy with it.

garrigue reviewed (approved):
Now it's 100% safe.

trefis commented:
Thank you!
